### ISSUE TITLE
Require consensuses

### ISSUE THREAD

**Issue by ajohnson-nrl** [privcount/privcount#27]
Enforce consensus at initialization and configuration

**Comment by teor2345** [comment]
Consensus among the SKs, with input from the DCs.

Agreement protocol:
* Takes a document as input
* TS sends the document to each DC and SK
* DCs send a document to each SK
* each SK checks that each DC listed in the document gave them the same document
* SKs send a document to each SK
* each SK checks that each SK listed in the document gave them the same document
* if either of these checks fail, tell the TS that initialisation or this particular configuration failed

Phases:
* Initialisation (once)
  * requires all nodes to participate, and verify parameters in Deployment Document
  * requires unanimous agreement (or actually, weak byzantine consensus) about the Deployment Document
  * agreement failure means the SK transmits an error to the TS and fails
* Configuration (repeatedly under TS control)
  * automated process, TS can trigger additional rounds
  * all other nodes can automatically verify Configuration Document
  * requires unanimous agreement (or actually, weak byzantine consensus) about the Configuration Document
  * requires other checks also pass:
    * must check delay from #22
    * must check noise amount from #9 
    * must check the sigma values respect the episilon / sigma constraints
      * input sigmas and sensitivity
      * use get_differentially_private_epsilon() to calculate epsilons
      * the sum of the epsilons is less than the epsilon value in the deployment document
  * agreement failure means the SK refuses this round, but may proceed with future rounds
    * sending another Configuration document tells SKs that think they had consensus that it failed elsewhere, and they should start again
* Execution
  * Setup
  * Collection
  * Aggregation

Deployment document:
* epsilon and delata
* sensitivity of the statistic
* reconfiguration time (#22)
* expected value & noise weight
  * depends on excess noise ratio, which depends on minimum DC number
* the protocol automatically enforces some constraints, but the set of verified information needs to be expansible for future changes:
  * minimum SK number (all SKs are currently required by the protocol)

Configuration document:
* start and end times
* counters
  * names
  * bins
  * sigma

**Comment by teor2345** [comment]
The consensus protocol needs to be documented in a doc/ directory, and a summary mentioned in the README.

**Comment by teor2345** [comment]
One consequence of this protocol is that each SK needs to know each other SK's public key.
There are a few ways I can think of to achieve this:
1. the TS includes the SK public keys in the Deployment Document it sends out to each DC, and each SK checks its own key is correct in the document (this implicitly occurs when it descrypts the document), or
2. the DCs adds the SK public keys to the Deployment Document they send out to each SK, and each SK checks its own key is in the document (this implicitly occurs when it descrypts the document), or
3. each SK is configured with each other SK's public key before initialisation.

@ajohnson-nrl, I prefer the final option, because the security proof is trivial. Are you happy to distribute all the SK public keys to each SK as part of the initial config?

**Comment by teor2345** [comment]
@ajohnson-nrl, are we concerned that the TS might perform an attack like:
1. drop the document sent from a DC to a SK
2. use the SK public key to encrypt a fake document

What about replay attacks:
1. keep the document sent from a DC to a SK in a previous round
2. drop the document sent from a DC to a SK in the current round
2. replay the previous round's document to the SK in the current round

**Comment by ajohnson-nrl** [comment]
Yes, all SKs need to "independently" verify the public signing keys of the other SKs, the TS, and the DCs. That is, they need to have some method outside the protocol to obtain keys that correspond to entities about which the SKs can make decisions regarding those entities' suitability for roles as SKs, DCs, and TS. This is required to prevent a man-in-the-middle attack (which would be particularly easy for the TS to perform). The DCs need to independently verify the keys of the SKs (they don't necessarily need to verify DC and TS keys because they can rely on the SKs to do that). The TS probably should independently verify the SK keys so that it's not fooled about whether or not its receiving valid status updates about a collection and final outputs.

We have just been putting this part of the process in the step during initialization at which the operator of the SK, DC, or TS accepts an initialization document, which contains the keys of all the parties. Perhaps there are ways of doing this that are less prone to operator error or laziness, however.

**Comment by ajohnson-nrl** [comment]
We are concerned that the TS might attempt to impersonate DCs (e.g. a MITM attack). That is why SKs must be sure that they have the correct DC keys before accepting the initialization document.

Replay attacks are another important concern. Avoiding them across configuration rounds (using the same initialization) is straightforward. Either the DCs and SKs can check to make sure that the start time is after the end time of the last successful collection (note that no assumption is needed about the accuracy of the local clocks) or a configuration round number can be added and checked to make sure it is increased by one over the last configuration round.

Avoiding replay attacks across initialization rounds seems a bit tricker in practice because we can't assume state is maintained at any party across rounds. One solution would be for each nonce SK to produce a nonce, send it to each of the other SKs, and then require that it be added to the initialization document that is later broadcast by the SKs to each other.

**Comment by teor2345** [comment]
Another solution for replay attacks is to include an expiry date in each document. I think we should do both that and a nonce.

**Comment by teor2345** [comment]
Un-tagging help wanted, because github now advertises issues with help wanted on the project page.
